### ISSUE TITLE
Read in and add commit messages

### ISSUE THREAD

**Issue by bockthom** [se-sic/coronet#180]
### Description

At the moment, we cannot access commit messages in `coronet`. However, as `coronet` builds on [codeface-extraction](https://github.com/se-passau/codeface-extraction), we could use the `commitMessages.list` file created by the commit message-extraction of `codeface-extraction`. Hence, we just need to read this file and somehow add it to the commit data.

### Ideas for the Implementation

There are, at least, two parts which need to be implemented:

**Part 1: Read `commitMessages.list`** 
Add a new function `read.commit.messages(data.path)` to `util-read.R`

**Part 2: Add a data structure which stores the commit messages to the `ProjectData` class and makes them accessible.**
Idea I: Add commit message as a new (optional) column to the commit data.
Idea II: Store commit message separately but provide a function which can join commit data and commit-messages data.

In addition, it might be useful to **preprocess the commit messages** already while reading them, that is, 
- retrieve the title of a commit message (i.e., first line)
- add correct line breaks (at the moment, there is no line break, but line breaks are represented by an "indentation" of four spaces.

**Comment by clhunsen** [comment]
I really like the idea to make data available in coronet and, thus, for project analysis through documented interfaces. Below you find my two cents regarding the implementation ideas.

> Idea I: Add commit message as a new (optional) column to the commit data.

If I understand this correctly, this idea would yield a solution that is analogous to the PaStA-related implementation: Add a `ProjectConf` attribute that configures whether the data is read from disk (`get.commit.messages`) and automatically added to the commit data. This implementation would be really consistent. :+1: 

To be more flexible, adding the comit-messages data to the commit data could be made optional – to prevent bloating the commit data. (Analogously, this could be implemented for PaStA and both commit and mail data.) Maybe, there should be a "global" configuration option that indicates whether orthogonal/addiational data sources are to be merged with the main data sources. :thinking: 

Regarding the `NetworkConf` class, I would not really like adding the data as edge attributes since the data would really bloat the `igraph` objects. But, to be flexible towards any use case, this should be an edge attribute that is, at least, disabled by default.

> Idea II: Store commit message separately but provide a function which can join commit data and commit-messages data.

This seems to be a subpart of Idea I, especially the flexible addition of commit messages to the commit data that I outlined above.

**Comment by bockthom** [comment]
> line breaks are represented by an "indentation" of four spaces.

As we had a discussion about whether the above stated "indentation" of four spaces in commit messages is generally valid, I have started some investigations on that. I just looked into our database and found out that there are actually *5 spaces* between  in-between two lines of what we assume to be a single line of a commit message. But this just a random observation or does this hold in general?

**tl;dr**:  In general, that is, the way our commit-message data is pre-processed, there actually is an indentation of *5 spaces*.

In the following, I will provide the detailed outcomes of my investigations for documentation reasons:

---

I started my investigations backwards, backtracing where our data come from. The `commitMessage.list` file, which we need to read in, is generated by [codeface-extraction](https://gibthub.com/se-passau/codeface-extraction), which in turn extracts the data from the [Codeface](https://gibthub.com/se-passau/codeface-extraction) database. So the question is: What does Codeface prior to writing the commit message into the database?

When looking into the source code of Codeface, we can see that Codeface also claims that there is an indentation of four spaces and also provides an example for that:
https://github.com/se-passau/codeface/blob/5b469fe1f633cda7e3601cdcb81919c05e596239/codeface/VCS.py#L806
https://github.com/se-passau/codeface/blob/5b469fe1f633cda7e3601cdcb81919c05e596239/codeface/VCS.py#L893

So the next question is: Where does Codeface's assumption come from? Why is the provided example generally valid? To answer this question, we need to find out what Codeface does to get the commit messages from the git repo. The simple answer is that Codeface just executes [`"git show --format=full --shortstat --numstat"`](https://github.com/se-passau/codeface/blob/5b469fe1f633cda7e3601cdcb81919c05e596239/codeface/VCS.py#L857) within its `_parseCommit` function. However, can we show that `git show` always indents commit messages by 4 spaces? Let's have a look at the implementation of git itself.

`git log` or `git show` internally use the function `pretty_print_commit` to print commits. At the beginning of this function, there is a statement [`int indent = 4;`](https://github.com/git/git/blob/54e85e7af1ac9e9a92888060d6811ae767fea1bc/pretty.c#L1942)—this sounds like what we are looking for—but is this really used to indent commit messages? So let's find out what's concretely going on in this function: https://github.com/git/git/blob/54e85e7af1ac9e9a92888060d6811ae767fea1bc/pretty.c#L1937-L2012
After preprocessing the header (`pp_header`), first the title line of the commit message is printed (`pp_title_line`) and then the remainder of the commit message is printed (`pp_remainder`) and, fortunately, the `indent` variable is passed as a parameter to the `pp_remainder` function. Within the `pp_remainder` function we can see that the `indent` parameter is actually used to indent every line of the commit message by calling [`pp_handle_indent(pp, sb, indent, line, linelen)`](https://github.com/git/git/blob/54e85e7af1ac9e9a92888060d6811ae767fea1bc/pretty.c#L1922) on each line of the commit message, which actually prints the indentation given by the `indent` parameter and also the whole line.

Hence, we now know that there is an indentation of 4 spaces at the beginning of each line of the commit message. But there still are the line breaks. So, wouldn't it be enough to just search for line breaks? The easy answer is: No, since Codeface splits the commit message by line breaks (https://github.com/se-passau/codeface/blob/5b469fe1f633cda7e3601cdcb81919c05e596239/codeface/VCS.py#L949) and joins the lines afterwards, using a single space as separator (https://github.com/se-passau/codeface/blob/5b469fe1f633cda7e3601cdcb81919c05e596239/codeface/commit.py#L159). So, combining this single space, which represents the line break, with the indentation of 4 spaces, we result in 5 spaces as line break :smile: 

:black_square_button:  quod erat demonstrandum 

---

To sum up, in the output we get from `git`, there is a line break followed by an indentation of 4 spaces. Codeface replaces the line break by an additional space, to increase readability. So, the general assumption for our implementation could be to search for 5 spaces in a row (as 4 spaces could just represent a tab) and replace them by a line break. However, this is just a heuristic, we cannot preclude that someone has deliberately added 5 spaces to his/her commit message :wink: 

(It was way more complicated to investigate where this indentation comes from, as it is nowhere specified in the documentation of `git`, but I am happy that I was able to find the corresponding code snippets in the implementation of `git` itself :smiley:)

**Comment by nlschn** [comment]
That is interesting, because in the sample files for busybox there only seem to be four spaces. So I wrote a prototype for the read function to deal with four spaces. I seem to have to change that though. 🤔

**Comment by bockthom** [comment]
> in the sample files for busybox there only seem to be four spaces
Really? Which files do you refer to? All files I had a look at had 5 spaces. Hence, I will check that again.

**Comment by bockthom** [comment]
Ah, I now understand what you may have meant @nlschn. There are only four spaces at the beginning of a commit message, i.e., in front of the "first line". But between the first and the second line, there are five spaces—an this should hold also for the remaining line breaks...

**Comment by nlschn** [comment]
`"cc8ed39b240180b58810784f844e253263594ac3";"    Initial revision"`
`4574675;"fbb39c83b69d6c4de943c0b7374000339635d13d";"    Stuff      -Erik"`

These are two lines from the`codeface-data/results/busybox_feature/feature/commitMessages.list` and at least they each begin with only four spaces.

**Comment by nlschn** [comment]
So do I get it right, that the first four spaces can be ignored generally?

**Comment by bockthom** [comment]
> So do I get it right, that the first four spaces can be ignored generally?

Yes, I think so.
